tests/unit/generator/lib.php

```diff
@@ -0,0 +1,164 @@
+            array_push($parts_created, $partid);
+        }
+
+        return $parts_created;
+    }
+    
+    /**
+     * Make a test Turnitin assignment module for use in various test cases.
+     * @param int $courseid - Moodle course ID
+     * @param string $modname - Module name (turnitintool or turnitintooltwo)
+     * @param int $assignmentid - Assignment id to which the coursemodule should be added
+     *
+     * @return  int $cm - id of the course module added
+     */
+    public static function make_test_module($courseid, $modname, $assignmentid)
+    {
+        global $DB;
+        // Set up a course module.
+        $module = $DB->get_record("modules", array("name" => $modname));
+        $coursemodule = new stdClass();
+        $coursemodule->course = $courseid;
+        $coursemodule->module = $module->id;
+        $coursemodule->added = time();
+        $coursemodule->instance = $assignmentid;
+        $coursemodule->section = 0;
+        $cmid = add_course_module($coursemodule);
+        
+        return $cmid;
+    }
+
+    /**
+     * Creates a moodle user and a corresponding entry in the turnitintooltwo_users table
+     * for the tii user specified
+     * @param  object $turnitintooltwo_user - turnitintooltwo user object
+     *
+     * @return  int $turnitintooltwo_user_id id of turnitintool user join (for use in get_record queries on turnitintooltwo_users table)
+     */
+    public static function join_test_user($turnitintooltwo_user)
+    {
+        $mdl_user = advanced_testcase::getDataGenerator()->create_user();
+        $tiiUserRecord = new stdClass();
+        $tiiUserRecord->userid = $mdl_user->id;
+        $tiiUserRecord->turnitin_uid = $turnitintooltwo_user->id;
+        $tiiUserRecord->user_agreement_accepted = 1;
+        $turnitintooltwo_user_id = $DB->insert_record('turnitintooltwo_users', $tiiUserRecord);
+        
+        return $turnitintooltwo_user_id;
+    }
+
+    /**
+     * Creates a number of test turnitintooltwo users, creates an equivalent moodle user for each, and handles the database association work.
+     *
+     * @param int $number_of_users
+     * @param array $roles - an array of strings, each of which should be 'learner' or 'instructor'.
+     * @return object $return - object of two arrays of equal length, one full of turnitintooltwo_user types and the other with ids for dbtable turnitintooltwo_users. The indices of these arrays DO align.
+     */
+    public function make_test_users($number_of_users, $roles)
+    {
+        $return = new stdClass();
+        $return->turnitintooltwo_users = array();
+        $return->joins = array();
+
+        for ($i=0; $i < $number_of_users; $i++) {
+            $new_user = new turnitintooltwo_user( $i+1, ($roles[$i] || 'instructor'), false, 'site', false );
+            array_push($return->turnitintooltwo_users, $new_user);
+        }
+
+        foreach ($turnitintooltwo_users as $turnitintooltwo_user) {
```

Comment:
It might be better to remove this `foreach` and move the logic to the `for` loop. The current setup is essentially doubling the overhead.